File: examples/serving/inference_csharp/SeqRecHelper.cs
using System;
using System.Collections.Generic;
using Microsoft.ML.OnnxRuntime.Tensors;
using System.Linq;
using System.IO;


namespace inference_csharp
{

    class Utilities
    {
        /// <summary>
        /// load user history and test set
        /// </summary>
        public static Data LoadData(string history_file, string feature_file, string test_file)
        {
            
            Data data = new Data();
            using (StreamReader sr = new StreamReader(history_file))
            {
                var line = sr.ReadLine();
                while (true)
                {
                    line = sr.ReadLine();
                    if (line == null)
                        break;

                    History_Node his_node = new History_Node();
                    string[] list = line.Trim().Split("\t");
                    his_node.user_id = Convert.ToInt64(list[0].Trim());
                    his_node.item_seq = list[1].Trim().Split(",").Select(x => Convert.ToInt64(x.Trim())).ToArray();
                    if (list.Length == 3)
                    {   
                        his_node.has_time = true;
                        his_node.time_seq = list[2].Trim().Split(",").Select(x => Convert.ToInt64(x.Trim())).ToArray();
                    }
                    else
                        {his_node.has_time = false;}
                    data.user_history[his_node.user_id] = his_node;
                }
            }

            if (GlobalVar.has_feature)
            {
                data.item_features = new Feature_Node[GlobalVar.n_items];
                data.item_features[0] = new Feature_Node();
                using (StreamReader sr = new StreamReader(feature_file))
                {
                    var line = sr.ReadLine();
                    while (true)
                    {
                        line = sr.ReadLine();
                        if (line == null)
                            break;

                        Feature_Node fea_node = new Feature_Node();
                        string[] list = line.Trim().Split("\t");
                        fea_node.item_id = Convert.ToInt64(list[0].Trim());
                        fea_node.item_feats = list[1].Trim().Split(",").Select(x => Convert.ToInt64(x.Trim())).ToArray();
                        data.item_features[fea_node.item_id] = fea_node;
                    }
                }

            }

            using (StreamReader sr = new StreamReader(test_file))
            {
                var line = sr.ReadLine();
                while (true)
                {
                    line = sr.ReadLine();
                    if (line == null)
                        break;

                    Test_Node t_node = new Test_Node();

                    string[] list = line.Trim().Split("\t");
                    t_node.user_id = Convert.ToInt64(list[0].Trim());
                    t_node.item_id = Convert.ToInt64(list[1].Trim());

                    data.test_set.Add(t_node);
                }
            }

            return data;
        }


        /// <summary>
        /// collate for one batch data, including item_seq, item_seq_len, time_seq(if exists)
        /// default pad left with 0
        /// currently only support 'autoregressive' history_mask_mode
        /// l and r are the left and right index in the test set corresponding to the batch
        /// </summary>
        public static Batch_Input collate_batch(Data data, int l, int r)
        {
            Batch_Input batch_input = new Batch_Input(r-l);
            for (int i=l; i<r; i++)
            {
                batch_input.user_id[i-l] = data.test_set[i].user_id;
                batch_input.item_id[i-l] = data.test_set[i].item_id;
                History_Node history = data.user_history[data.test_set[i].user_id];
                long item_id = data.test_set[i].item_id;
                int history_len = history.item_seq.Length-1;
                for (; history_len>=0; history_len--)
                {
                    if (history.item_seq[history_len] == item_id)
                        {break;}
                }
                if (history_len == -1)
                    {history_len = history.item_seq.Length;}
                batch_input.item_seq_len[i-l] = Math.Min(history_len, GlobalVar.max_seq_len);
                for ( int j=history_len-1, k=GlobalVar.max_seq_len-1 ; j>=Math.Max(0, history_len-GlobalVar.max_seq_len); j--, k-- )
                {
                    batch_input.item_seq[i-l, k] = history.item_seq[j];
                    if (history.has_time)
                        {batch_input.time_seq[i-l, k] = history.time_seq[j];}
                }
            }
            if (GlobalVar.has_feature)
            {
                batch_input.item_features = new DenseTensor<long>(new[] { r-l, GlobalVar.n_features });
                for (int i=l; i<r; i++)
                {
                    for (int j=0; j<GlobalVar.n_features; j++)
                    {
                        batch_input.item_features[i-l, j] = data.item_features[data.test_set[i].item_id].item_feats[j];
                    }
                }
                batch_input.item_seq_features = new DenseTensor<long>(new[] { r-l, GlobalVar.max_seq_len, GlobalVar.n_features });
                // based on batch_input.item_seq
                for (int i=l; i<r; i++)
                {
                    for (int j=0; j<GlobalVar.max_seq_len; j++)
                    {
                        for (int k=0; k<GlobalVar.n_features; k++)
                        {
                            batch_input.item_seq_features[i-l, j, k] = data.item_features[batch_input.item_seq[i-l, j]].item_feats[k];
                        }
                    }
                }
            }
            return batch_input;
        }
    }
}

File: examples/serving/inference_java/src/main/java/com/example/SeqRecHelper.java
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

package com.example;
import java.io.BufferedReader;  
import java.io.FileReader;  
import java.io.IOException;  
import java.util.*;  
  
public class SeqRecHelper {  
  
    public static DataNode.Data loadData(String historyFile, String featureFile, String testFile) {  
        DataNode.Data data = new DataNode.Data();  
  
        try (BufferedReader br = new BufferedReader(new FileReader(historyFile))) {  
            String line;  
            line = br.readLine();
            while ((line = br.readLine()) != null) {  
                DataNode.HistoryNode historyNode = new DataNode.HistoryNode();  
                String[] list = line.trim().split("\t");  
                historyNode.userId = Long.parseLong(list[0].trim());  
                historyNode.itemSeq = Arrays.stream(list[1].trim().split(",")).mapToLong(Long::parseLong).toArray();  
                if (list.length == 3) {  
                    historyNode.hasTime = true;  
                    historyNode.timeSeq = Arrays.stream(list[2].trim().split(",")).mapToLong(Long::parseLong).toArray();
                } else {  
                    historyNode.hasTime = false;  
                }  
                data.userHistory[(int) historyNode.userId] = historyNode;  
            }  
        } catch (IOException e) {  
            e.printStackTrace();  
        }  
  
        if (Main.hasFeature) {  
            data.itemFeatures = new DataNode.FeatureNode[Main.nItems];  
            data.itemFeatures[0] = new DataNode.FeatureNode();  
            try (BufferedReader br = new BufferedReader(new FileReader(featureFile))) {  
                String line;  
                line = br.readLine();
                while ((line = br.readLine()) != null) {  
                    DataNode.FeatureNode featureNode = new DataNode.FeatureNode();  
                    String[] list = line.trim().split("\t");  
                    featureNode.itemId = Long.parseLong(list[0].trim());  
                    featureNode.itemFeats = Arrays.stream(list[1].trim().split(",")).mapToLong(Long::parseLong).toArray();  
                    data.itemFeatures[(int) featureNode.itemId] = featureNode;  
                }  
            } catch (IOException e) {  
                e.printStackTrace();  
            }  
        }  
  
        data.testSet = new ArrayList<>();  
        try (BufferedReader br = new BufferedReader(new FileReader(testFile))) {  
            String line;  
            line = br.readLine();
            while ((line = br.readLine()) != null) {  
                DataNode.TestNode testNode = new DataNode.TestNode();  
                String[] list = line.trim().split("\t");  
                testNode.userId = Long.parseLong(list[0].trim());  
                testNode.itemId = Long.parseLong(list[1].trim());  
                data.testSet.add(testNode);  
            }  
        } catch (IOException e) {  
            e.printStackTrace();  
        }  
  
        return data;  
    }  
  
    public static DataNode.BatchInput collateBatch(DataNode.Data data, int l, int r) {  
        DataNode.BatchInput batchInput = new DataNode.BatchInput(r - l);  
        for (int i = l; i < r; i++) {  
            batchInput.userId[i - l] = data.testSet.get(i).userId;  
            batchInput.itemId[i - l] = data.testSet.get(i).itemId;  
            DataNode.HistoryNode history = data.userHistory[(int) data.testSet.get(i).userId];  
            long itemId = data.testSet.get(i).itemId;  
            int historyLen = history.itemSeq.length - 1;  
            for (; historyLen >= 0; historyLen--) {  
                if (history.itemSeq[historyLen] == itemId) {  
                    break;  
                }  
            }  
            if (historyLen == -1) {  
                historyLen = history.itemSeq.length;
            }
            batchInput.itemSeqLen[i - l] = Math.min(historyLen, Main.maxSeqLen);  
            for (int j = historyLen - 1, k = Main.maxSeqLen - 1; j >= Math.max(0, historyLen - Main.maxSeqLen); j--, k--) {  
                batchInput.itemSeq[i - l][k] = history.itemSeq[j];  
                if (history.hasTime) {  
                    batchInput.timeSeq[i - l][k] = history.timeSeq[j];  
                }  
            }  
        }  
        if (Main.hasFeature) {  
            batchInput.itemFeatures = new long[r - l][Main.nFeatures];  
            for (int i = l; i < r; i++) {  
                for (int j = 0; j < Main.nFeatures; j++) {  
                    batchInput.itemFeatures[i - l][j] = data.itemFeatures[(int) data.testSet.get(i).itemId].itemFeats[j];  
                }  
            }  
            batchInput.itemSeqFeatures = new long[r - l][Main.maxSeqLen][Main.nFeatures];  
            for (int i = l; i < r; i++) {  
                for (int j = 0; j < Main.maxSeqLen; j++) {  
                    for (int k = 0; k < Main.nFeatures; k++) {  
                        batchInput.itemSeqFeatures[i - l][j][k] = data.itemFeatures[(int) batchInput.itemSeq[i - l][j]].itemFeats[k];  
                    }  
                }  
            }  
        }  
        return batchInput;  
    }  
}  
